File: idakonyvtar/src/main/java/eu/eyan/idakonyvtar/controller/BookController.java
package eu.eyan.idakonyvtar.controller;

import static com.google.common.collect.Lists.newArrayList;

import java.awt.Component;
import java.awt.Window;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.beans.PropertyChangeEvent;
import java.util.List;
import java.util.stream.Collectors;

import javax.swing.ImageIcon;
import javax.swing.JComboBox;
import javax.swing.JOptionPane;
import javax.swing.JTextField;
import javax.swing.SwingUtilities;

import org.fest.util.Objects;
import org.jdesktop.swingx.autocomplete.AutoCompleteDecorator;

import com.google.common.base.Joiner;
import com.google.common.io.Resources;
import com.jgoodies.binding.adapter.Bindings;
import com.jgoodies.binding.adapter.ComboBoxAdapter;

import eu.eyan.idakonyvtar.controller.input.BookControllerInput;
import eu.eyan.idakonyvtar.model.Book;
import eu.eyan.idakonyvtar.model.BookFieldValueModel;
import eu.eyan.idakonyvtar.model.ColumnKonfiguration.ColumnConfigurations;
import eu.eyan.idakonyvtar.oszk.Marc;
import eu.eyan.idakonyvtar.oszk.OszkKereso;
import eu.eyan.idakonyvtar.oszk.OszkKeresoException;
import eu.eyan.idakonyvtar.util.BookHelper;
import eu.eyan.idakonyvtar.view.BookView;
import eu.eyan.idakonyvtar.view.MultiField;
import eu.eyan.idakonyvtar.view.MultiFieldJComboBox;
import eu.eyan.idakonyvtar.view.MultiFieldJTextField;

public class BookController implements IDialogController<BookControllerInput, Book>
{
    private BookView view = new BookView();
    private BookControllerInput model;
    private List<Window> resizeListeners = newArrayList();

    @Override
    public Component getView()
    {
        return view.getComponent();
    }

    @Override
    public String getTitle()
    {
        if (model.getColumns().indexOf("Szerző") >= 0)
        {
            return "Book adatainak szerkesztése - " + model.getBook().getValue(model.getColumns().indexOf("Szerző"));
        }
        else
        {
            return "Book adatainak szerkesztése";
        }
    }

    @Override
    public void initData(BookControllerInput model)
    {
        this.model = model;
        view.setColumns(model.getColumns());
        view.setIsbnEnabled(model.isIsbnEnabled());
        view.setColumnConfiguration(model.getColumnConfiguration());
    }

    @Override
    public void initBindings()
    {
        initFieldsActionBindings();
        view.getIsbnText().addActionListener(isbnSearch());
    }

    private void initFieldsActionBindings()
    {
        for (int columnIndex = 0; columnIndex < model.getColumns().size(); columnIndex++)
        {
            String columnName = model.getColumns().get(columnIndex);
            boolean autoComplete = model.getColumnConfiguration().isTrue(columnName, ColumnConfigurations.AUTOCOMPLETE);
            boolean multi = model.getColumnConfiguration().isTrue(columnName, ColumnConfigurations.MULTIFIELD);
            if (autoComplete)
            {
                List<String> columnList = BookHelper.getColumnList(model.getBookList(), columnIndex);
                if (multi)
                {
                    MultiFieldJComboBox mmcombo = (MultiFieldJComboBox) view.getEditors().get(columnIndex);
                    mmcombo.setAutoCompleteList(columnList);
                    multimezőBind(mmcombo, new BookFieldValueModel(columnIndex, model.getBook()));
                }
                else
                {
                    JComboBox<?> comboBox = (JComboBox<?>) view.getEditors().get(columnIndex);
                    Bindings.bind(comboBox, new ComboBoxAdapter<String>(columnList, new BookFieldValueModel(columnIndex, model.getBook())));
                    AutoCompleteDecorator.decorate(comboBox);
                }
            }
            else
            {
                if (multi)
                {
                    MultiFieldJTextField mmc = (MultiFieldJTextField) view.getEditors().get(columnIndex);
                    multimezőBind(mmc, new BookFieldValueModel(columnIndex, model.getBook()));
                }
                else
                {
                    Bindings.bind((JTextField) view.getEditors().get(columnIndex), new BookFieldValueModel(columnIndex, model.getBook()));
                }
            }
        }
    }

    private void multimezőBind(final MultiField<String, ?> mmc, final BookFieldValueModel bookFieldValueModel)
    {
        bookFieldValueModel.addValueChangeListener((PropertyChangeEvent propertyChangeEvent) -> {
            if (!Objects.areEqual(propertyChangeEvent.getNewValue(), propertyChangeEvent.getOldValue()))
            {
                mmc.setValues(getMultiFieldList((String) propertyChangeEvent.getNewValue()));
            }
        });
        mmc.setValues(getMultiFieldList((String) bookFieldValueModel.getValue()));

        mmc.addPropertyChangeListener((PropertyChangeEvent propertyChangeEvent) -> {
            bookFieldValueModel.setValue(Joiner.on(BookHelper.LISTA_SEPARATOR).skipNulls().join(mmc.getValues()));
        });
    }

    private static List<String> getMultiFieldList(String value)
    {
        String[] strings = (value).split(BookHelper.LISTA_SEPARATOR_REGEX);
        List<String> list = newArrayList(strings).stream().filter((String s) -> {
            return !s.isEmpty();
        }).collect(Collectors.toList());
        return list;
    }

    private ActionListener isbnSearch()
    {
        return new ActionListener()
        {
            @Override
            public void actionPerformed(ActionEvent e)
            {
                if (e.getSource() == view.getIsbnText())
                {
                    System.out.println("ISBN Action: " + view.getIsbnText().getText());
                    view.getIsbnText().selectAll();
                    view.getIsbnSearchLabel().setText("Keresés");
                    view.getIsbnSearchLabel().setIcon(new ImageIcon(Resources.getResource("icons/search.gif")));
                    view.getEditors().forEach(component -> component.setEnabled(false));
                    // TODO Asynchron
                    SwingUtilities.invokeLater(new Runnable()
                    {
                        @Override
                        public void run()
                        {
                            List<Marc> marcsToIsbn;
                            try
                            {
                                marcsToIsbn = OszkKereso.getMarcsToIsbn(view.getIsbnText().getText().replaceAll("ö", "0"));
                                prozessIsbnData(marcsToIsbn);
                            }
                            catch (OszkKeresoException e)
                            {
                                // FIXME: itt fontos a naplózás
                                view.getIsbnSearchLabel().setText("Nincs találat");
                                view.getIsbnSearchLabel().setIcon(new ImageIcon(Resources.getResource("icons/error.gif")));
                            }
                            finally
                            {
                                view.getEditors().forEach(component -> component.setEnabled(true));
                                fireResizeEvent();
                            }
                        }
                    });
                    fireResizeEvent();
                }
            }
        };
    }

    private void prozessIsbnData(List<Marc> marcsToIsbn)
    {
        for (String column : model.getColumns())
        {
            String columnValue = "";
            List<Marc> marcCodesToColumns;
            try
            {
                marcCodesToColumns = model.getColumnConfiguration().getMarcCodes(column);
                for (Marc marc : marcsToIsbn)
                {
                    for (Marc columnMarc : marcCodesToColumns)
                    {
                        if (isMarcsApply(marc, columnMarc))
                        {
                            columnValue += columnValue.equals("") ? marc.getValue() : ", " + marc.getValue();
                        }
                    }
                }
                model.getBook().setValue(model.getColumns().indexOf(column), columnValue);
            }
            catch (Exception e)
            {
                e.printStackTrace();
                JOptionPane.showMessageDialog(null, e.getLocalizedMessage());
            }
        }
    }

    private boolean isMarcsApply(Marc pontosMarc, Marc pontatlanMarc)
    {
        if (pontosMarc == null || pontatlanMarc == null || pontosMarc.getMarc1() == null || pontatlanMarc.getMarc1() == null)
        {
            return false;
        }
        if (pontosMarc.getMarc1().equalsIgnoreCase(pontatlanMarc.getMarc1()))
        {
            if (pontatlanMarc.getMarc2().equals("") || pontosMarc.getMarc2().equalsIgnoreCase(pontatlanMarc.getMarc2()))
            {
                if (pontatlanMarc.getMarc3().equals("") || pontosMarc.getMarc3().equalsIgnoreCase(pontatlanMarc.getMarc3()))
                {
                    return true;
                }
            }
        }
        return false;
    }

    @Override
    public void onOk()
    {
    }

    @Override
    public void onCancel()
    {
    }

    @Override
    public Book getOutput()
    {
        return model.getBook();
    }

    @Override
    public Component getComponentForFocus()
    {
        return view.getIsbnText();
    }

    @Override
    public void addResizeListener(Window window)
    {
        this.resizeListeners.add(window);
    }

    private void fireResizeEvent()
    {
        for (Window window : resizeListeners)
        {
            window.pack();
        }
    }
}

File: idakonyvtar/src/main/scala/eu/eyan/idakonyvtar/controller/BookController.scala
package eu.eyan.idakonyvtar.controller;

import java.awt.Component
import java.awt.Image
import java.awt.Window
import java.awt.event.ActionListener
import java.beans.PropertyChangeEvent
import java.beans.PropertyChangeListener
import java.io.File

import scala.collection.JavaConversions.asScalaBuffer
import scala.collection.JavaConversions.seqAsJavaList
import scala.collection.mutable.MutableList

import com.google.common.collect.Lists.newArrayList
import com.google.common.io.Resources
import com.jgoodies.binding.adapter.Bindings
import com.jgoodies.forms.layout.FormLayout
import com.jgoodies.forms.layout.RowSpec

import eu.eyan.idakonyvtar.model.Book
import eu.eyan.idakonyvtar.model.BookFieldValueModel
import eu.eyan.idakonyvtar.oszk.Marc
import eu.eyan.idakonyvtar.oszk.OszkKereso
import eu.eyan.idakonyvtar.oszk.OszkKeresoException
import eu.eyan.idakonyvtar.util.LibraryException
import eu.eyan.idakonyvtar.util.WebCam
import eu.eyan.idakonyvtar.view.MultiField
import eu.eyan.idakonyvtar.view.MultiFieldAutocomplete
import eu.eyan.idakonyvtar.view.MultiFieldJTextField
import eu.eyan.log.Log
import eu.eyan.util.awt.AwtHelper
import eu.eyan.util.awt.ComponentPlus.ComponentPlusImplicit
import eu.eyan.util.string.StringPlus.StringPlusImplicit
import eu.eyan.util.swing.JButtonPlus.JButtonImplicit
import eu.eyan.util.swing.JLabelPlus.JLabelImplicit
import eu.eyan.util.swing.JPanelWithFrameLayout
import eu.eyan.util.swing.JTextFieldAutocomplete
import eu.eyan.util.swing.SwingPlus
import javax.imageio.ImageIO
import javax.swing.ImageIcon
import javax.swing.JButton
import javax.swing.JLabel
import javax.swing.JOptionPane
import javax.swing.JPanel
import javax.swing.JTextField
import eu.eyan.idakonyvtar.text.TechnicalTextsIda._
import eu.eyan.util.swing.WithComponent
import eu.eyan.util.swing.JTextFieldPlus.JTextFieldPlusImplicit
import eu.eyan.idakonyvtar.model.BookField
import eu.eyan.idakonyvtar.util.ExcelHandler.FieldConfiguration

object BookController {
  def listForAutocomplete(bookList: Seq[Book], field: BookField) = bookList 
    .map(_.getValue(field)) // get the values of the column
    .filter(_ != null) // only not nulls
    .map(s => if (s.contains(MULTIFIELD_SEPARATOR)) s.split(MULTIFIELS_SEPARATOR_REGEX) else Array(s)) // get all values if multifield
    .flatten // take the whole list
    .++:(List("")) // empty is always the default option
    .map(_.trim)
    // .distinct //do distinct in ac
    //      .sortWith((s1: String, s2: String) => COLLATOR.compare(s1, s2) < 0) //autocomplete does sorting
    .toList
}

trait IsbnSetting
case object NO_ISBN extends IsbnSetting 
case object WITH_ISBN extends IsbnSetting
  
  
class BookController(
  private val book:                Book,
  private val fields:              List[BookField], 
  private val bookList:            List[Book],
  private val isbnEnabled:         IsbnSetting, //TODO make two applies for editing and for newBook
  private val loadedFile:          File) {

  def getComponent = view

  private val TEXTFIELD_DEFAULT_SIZE = 20
  private val editors: MutableList[Component] = MutableList()
  private val isbnSearchLabel = new JLabel()
  private val isbnText = new JTextField()
  private val webcamPanel = new JPanelWithFrameLayout
  private val resizeListeners: java.util.List[Window] = newArrayList()

  val fieldsPanel = new JPanelWithFrameLayout().withSeparators.newColumn.newColumn("pref:grow")

  val picturePanel = new JPanelWithFrameLayout().withSeparators.newColumn("320px")

  val view = new JPanelWithFrameLayout().withSeparators
    .newColumn("f:p:g ").addFluent(fieldsPanel)
    .newColumn("f:320px").addFluent(picturePanel)
    .newColumn("f:320px").addFluent(webcamPanel)

  startWebcam

  if (isbnEnabled == WITH_ISBN) {
    fieldsPanel.newRow.span.addSeparatorWithTitle("Isbn")
    fieldsPanel.newRow
    fieldsPanel.add(isbnSearchLabel.name(ISBN_LABEL))
    fieldsPanel.nextColumn.add(isbnText.name(ISBN_TEXT))

    isbnText onHierarchyChanged isbnText.requestFocusInWindow
    isbnText onActionPerformed isbnSearch
  }

  fieldsPanel.newRow.span.addSeparatorWithTitle("Adatok")

  for { fieldIndex <- 0 until fields.size } {
    val field = fields(fieldIndex)
    Log.debug(s"column $field")
    fieldsPanel.newRow.addLabel(field.fieldName)

    val isMultiEditorField = field.isMulti
    val isAutocompleteField = field.isAutocomplete
    val isPictureField = field.isPicture

    val editor: Component =
      if (isPictureField) {
        //TODO WTF spagetti:
        picturePanel.newRow.addLabel(field.fieldName)
        val imageLabel = picturePanel.newRow.addLabel("").name("look" + field)

        val imgNameAndBtn = JPanelWithFrameLayout()
        val textField = imgNameAndBtn.newColumn.addTextField("", TEXTFIELD_DEFAULT_SIZE).name("picturePath")
        val button = imgNameAndBtn.newColumn.addButton("katt").name("click")

        def refreshImage = imageLabel.setIcon(new ImageIcon(book.getImage(field).get.getScaledInstance(320, 240, Image.SCALE_DEFAULT)))
        if (book.getImage(field).nonEmpty) refreshImage
        button.onClicked({
          book.setImage(field)(WebCam.getImage)
          book.setValue(field)("")
          refreshImage
        })

        bindTextField(textField, new BookFieldValueModel(field, book))
        imgNameAndBtn
      } else if (isAutocompleteField)
        if (isMultiEditorField) multiFieldBind(new MultiFieldAutocomplete(field.fieldName, "Autocomplete", "Nincs találat").setAutoCompleteList(BookController.listForAutocomplete(bookList, field)), new BookFieldValueModel(field, book))
        else bindTextField(new JTextFieldAutocomplete().setHintText("Autocomplete").setAutocompleteList(BookController.listForAutocomplete(bookList, field)), new BookFieldValueModel(field, book))
      else if (isMultiEditorField) multiFieldBind(new MultiFieldJTextField(field.fieldName), new BookFieldValueModel(field, book))
      else bindTextField(new JTextField(TEXTFIELD_DEFAULT_SIZE), new BookFieldValueModel(field, book))

    editor.setName(field.fieldName)
    fieldsPanel.nextColumn.add(editor)
    editors += editor
  }

  private def bindTextField(tf: JTextField, bookFieldValueModel: BookFieldValueModel) = {
    Bindings.bind(tf, bookFieldValueModel)
    tf
  }

  private def multiFieldBind(mmc: MultiField[String, _], bookFieldValueModel: BookFieldValueModel) = {
    Log.debug(mmc.getName + " " + bookFieldValueModel)
    bookFieldValueModel.addValueChangeListener(new PropertyChangeListener {
      override def propertyChange(evt: PropertyChangeEvent) =
        if (evt.getNewValue() != evt.getOldValue()) {
          Log.debug("Listener: " + evt.getOldValue + " -> " + evt.getNewValue)
          mmc.setValues(getMultiFieldList(evt.getNewValue().asInstanceOf[String]))
        }
    })

    mmc.setValues(getMultiFieldList(bookFieldValueModel.getValue()))

    mmc.addPropertyChangeListener(new PropertyChangeListener {
      override def propertyChange(evt: PropertyChangeEvent) =
        bookFieldValueModel.setValue(mmc.getValues().filter(_ != null).mkString(MULTIFIELD_SEPARATOR))
    })
    mmc
  }

  private def isbnSearch() = {
    isbnText.selectAll()
    isbnSearchLabel.setText("Keresés")
    isbnSearchLabel.setIcon(new ImageIcon(Resources.getResource("icons/search.gif")))
    editors.foreach(_.setEnabled(false))

    SwingPlus.invokeLater {
      {
        try {
          val marcsToIsbn = OszkKereso.getMarcsToIsbn(isbnText.getText().replaceAll("ö", "0"))
          prozessIsbnData(marcsToIsbn)
        } catch {
          case e: OszkKeresoException =>
            Log.error(e)
            isbnSearchLabel.setText("Nincs találat")
            isbnSearchLabel.setIcon(new ImageIcon(Resources.getResource("icons/error.gif")))
        } finally {
          editors.foreach(_.setEnabled(true))
        }
      }
    }
  }

  private def prozessIsbnData(marcsFromOszk: List[Marc]) =
    fields.foreach(field => {
      try {
        val marcCodesFromColumns = field.marcCodes
        val values = for {
          marcFromOszk <- marcsFromOszk
          marcFromColumn <- marcCodesFromColumns if (isMarcsApply(marcFromOszk, marcFromColumn))
        } yield marcFromOszk.value
        Log.info("BookController.prozessIsbnData " + values.mkString("\r\n    "))
        book.setValue(field)(values.mkString(", "))
      } catch {
        case e: Exception =>
          e.printStackTrace()
          JOptionPane.showMessageDialog(null, e.getLocalizedMessage())
      }
    })

  private def isMarcsApply(marcFromOszk: Marc, marcFromColumn: Marc) = {
    // Log.debug("marcFromOszk: " + marcFromOszk + " =?= marcFromColumn:" + marcFromColumn)
    if (marcFromOszk == null || marcFromColumn == null || marcFromOszk.marc1 == null || marcFromColumn.marc1 == null)
      false
    else marcFromOszk.marc1.equalsIgnoreCase(marcFromColumn.marc1) &&
      (marcFromColumn.marc2.equals("") || marcFromOszk.marc2.equalsIgnoreCase(marcFromColumn.marc2)) &&
      (marcFromColumn.marc3.equals("") || marcFromOszk.marc3.equalsIgnoreCase(marcFromColumn.marc3))
  }

  private def startWebcam = {
    Log.info("start webcam")
    val webcam = WebCam.startWebcam
    if (webcam.nonEmpty) {
      webcamPanel.newRow.add(webcam.get.picker)
      webcamPanel.newRow.add(webcam.get.panel)
    } else webcamPanel.add(new JLabel("No Webcam"))
  }

  private def getMultiFieldList(value: String): List[String] = value.split(MULTIFIELS_SEPARATOR_REGEX).filter(!_.isEmpty()).toList
}